From f125a460ea6eb869da6dcdd1bb95b2d892e2286c Mon Sep 17 00:00:00 2001
From: Akshat Kumar <akshat.kumar02@st.com>
Date: Fri, 22 Sep 2023 16:48:33 +0530
Subject: [PATCH] OPTEE-OS-STM32MP: 3.12.0-stm32mp-r2.3-rc1

---
 core/drivers/stm32_rng.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index e95bde8ad..05fccd837 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -103,9 +103,22 @@ TEE_Result stm32_rng_read_raw(vaddr_t rng_base, uint8_t *out, size_t *size)
 
 		/* RNG is ready: read up to 4 32bit words */
 		while (len) {
-			uint32_t data32 = io_read32(rng_base + RNG_DR);
+			uint32_t data32 = 0;
 			size_t sz = MIN(len, sizeof(uint32_t));
 
+			/*
+			 * Late seed error case: DR being 0 is an error
+			 * status.
+			 */
+			if (!(io_read32(rng_base + RNG_SR) & RNG_SR_DRDY))
+				break;
+			data32 = io_read32(rng_base + RNG_DR);
+
+			if (!data32) {
+				conceal_seed_error(rng_base);
+				return TEE_ERROR_NO_DATA;
+			}
+
 			memcpy(buf, &data32, sz);
 			buf += sz;
 			len -= sz;
-- 
2.25.1

