From 00d191a63ebc27ba5bc9f796f6dc7843e4a8d379 Mon Sep 17 00:00:00 2001
From: christophe montaud <christophe.montaud@st.com>
Date: Wed, 16 May 2018 13:50:17 +0200
Subject: [PATCH] Enable hardware watchdog inside systemd

---
 src/core/system.conf | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/core/system.conf.in b/src/core/system.conf.in
index f0a59a7..e5f5aaa 100644
--- a/src/core/system.conf.in
+++ b/src/core/system.conf.in
@@ -24,8 +24,8 @@
 #CtrlAltDelBurstAction=reboot-force
 #CPUAffinity=1 2
 #JoinControllers=cpu,cpuacct net_cls,net_prio
-#RuntimeWatchdogSec=0
-#ShutdownWatchdogSec=10min
+RuntimeWatchdogSec=30
+ShutdownWatchdogSec=2min
 #CapabilityBoundingSet=
 #NoNewPrivileges=no
 #SystemCallArchitectures=
-- 
2.7.4

