From 3f772653dee358ad80a23b0936788734f1693736 Mon Sep 17 00:00:00 2001
From: Romuald Jeanne <romuald.jeanne@st.com>
Date: Mon, 5 Jun 2023 14:17:30 +0200
Subject: [PATCH 22/24] v6.1-stm32mp-r1 MACHINE

Signed-off-by: Romuald Jeanne <romuald.jeanne@st.com>
---
 arch/arm/mach-stm32/board-dt.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/arm/mach-stm32/board-dt.c b/arch/arm/mach-stm32/board-dt.c
index 2ccaa11aaa56..9ff06f2fcbf4 100644
--- a/arch/arm/mach-stm32/board-dt.c
+++ b/arch/arm/mach-stm32/board-dt.c
@@ -21,6 +21,8 @@ static const char *const stm32_compat[] __initconst = {
 	"st,stm32mp131",
 	"st,stm32mp133",
 	"st,stm32mp135",
+	"st,stm32mp151",
+	"st,stm32mp153",
 	"st,stm32mp157",
 	NULL
 };
-- 
2.17.1

