From 31b053d77837fe81d1f6ed3af37a634a88bb173b Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
Date: Tue, 18 Oct 2022 11:42:52 +0200
Subject: [PATCH] LD: not track warning as error and no-warn-rwx-segment

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
---
 Makefile | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 35644b832..345d0c2c9 100644
--- a/Makefile
+++ b/Makefile
@@ -445,8 +445,10 @@ TF_LDFLAGS		+=	$(subst --,-Xlinker --,$(TF_LDFLAGS_$(ARCH)))
 
 # LD = gcc-ld (ld) or llvm-ld (ld.lld) or other
 else
-TF_LDFLAGS		+=	--fatal-warnings -O1
+#TF_LDFLAGS		+=	--fatal-warnings -O1
+TF_LDFLAGS		+=	 -O1
 TF_LDFLAGS		+=	--gc-sections
+TF_LDFLAGS		+=	--no-warn-rwx-segment
 # ld.lld doesn't recognize the errata flags,
 # therefore don't add those in that case
 ifeq ($(findstring ld.lld,$(notdir $(LD))),)
-- 
2.25.1

