From 1e8a2ee2730867c509209f6f8e3cd356c9729d4e Mon Sep 17 00:00:00 2001
From: Jean-Philippe ROMAIN <jean-philippe.romain@st.com>
Date: Mon, 28 Oct 2019 15:55:53 +0100
Subject: [PATCH 1/2] SDK: Add relative path for cmsis-svd


diff --git a/gdb-svd.py b/gdb-svd.py
index 0586755..9df9ac8 100644
--- a/gdb-svd.py
+++ b/gdb-svd.py
@@ -17,6 +17,9 @@
 # You should have received a copy of the GNU General Public License
 # along with svd-tools.  If not, see <https://www.gnu.org/licenses/>.
 
+import os
+sys.path.insert(0, os.path.dirname(__file__) + "/../cmsis-svd")
+
 import gdb
 from terminaltables import AsciiTable
 from cmsis_svd.parser import SVDParser
-- 
2.7.4

