From 65ad7ed110b25b333679e5084fb82db0d5a15a6c Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@st.com>
Date: Fri, 10 Apr 2020 14:50:46 +0200
Subject: [PATCH 22/23] ARM-stm32mp1-r1-POWER

---
 kernel/power/suspend.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/kernel/power/suspend.c b/kernel/power/suspend.c
index 27f149f5d..4581702e8 100644
--- a/kernel/power/suspend.c
+++ b/kernel/power/suspend.c
@@ -34,7 +34,6 @@
 #include "power.h"
 
 const char * const pm_labels[] = {
-	[PM_SUSPEND_TO_IDLE] = "freeze",
 	[PM_SUSPEND_STANDBY] = "standby",
 	[PM_SUSPEND_MEM] = "mem",
 };
-- 
2.17.1

