From 49a7d9003e2c89aa252d3ccccb16d4ab22a90ef6 Mon Sep 17 00:00:00 2001
From: Lionel Vitte <lionel.vitte@st.com>
Date: Thu, 14 Oct 2021 16:51:52 +0200
Subject: [PATCH 20/23] ARM 5.10.61-stm32mp1-r2 MISC

---
 drivers/opp/core.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/opp/core.c b/drivers/opp/core.c
index 903b465c8..dbcce5de9 100644
--- a/drivers/opp/core.c
+++ b/drivers/opp/core.c
@@ -1608,9 +1608,13 @@ struct opp_table *dev_pm_opp_set_supported_hw(struct device *dev,
 	struct opp_table *opp_table;
 
 	opp_table = dev_pm_opp_get_opp_table(dev);
-	if (IS_ERR(opp_table))
+
+	if (PTR_ERR(opp_table) == -EPROBE_DEFER)
 		return opp_table;
 
+	if (!opp_table)
+		return ERR_PTR(-ENOMEM);
+
 	/* Make sure there are no concurrent readers while updating opp_table */
 	WARN_ON(!list_empty(&opp_table->opp_list));
 
-- 
2.17.1

