From 2355a925068a138b888e04410aba515219ce7b38 Mon Sep 17 00:00:00 2001
From: Jean-Philippe ROMAIN <jean-philippe.romain@st.com>
Date: Wed, 30 Jan 2019 17:45:00 +0100
Subject: [PATCH] Install bpf files only if DKMS is enable


diff --git a/driver/bpf/CMakeLists.txt b/driver/bpf/CMakeLists.txt
index e8ae7b6..6208ecc 100644
--- a/driver/bpf/CMakeLists.txt
+++ b/driver/bpf/CMakeLists.txt
@@ -8,6 +8,7 @@ if(BUILD_BPF)
 		VERBATIM)
 endif()
 
+if(ENABLE_DKMS)
 install(FILES
 	bpf_helpers.h
 	filler_helpers.h
@@ -21,3 +22,4 @@ install(FILES
 	types.h
 	DESTINATION "src/${PACKAGE_NAME}-${PROBE_VERSION}/bpf"
 	COMPONENT agent)
+endif()
-- 
2.7.4

