From 2b0ce07da166390395e4def20f32c6eb8b22053b Mon Sep 17 00:00:00 2001
From: Akshat Kumar <akshat.kumar02@st.com>
Date: Mon, 25 Sep 2023 21:43:07 +0530
Subject: [PATCH 21/23] ARM 5.10.188-stm32mp1-r2.3-rc1 MISC

---
 drivers/opp/core.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/opp/core.c b/drivers/opp/core.c
index 7ed605ffb..06f808e27 100644
--- a/drivers/opp/core.c
+++ b/drivers/opp/core.c
@@ -1608,9 +1608,13 @@ struct opp_table *dev_pm_opp_set_supported_hw(struct device *dev,
 	struct opp_table *opp_table;
 
 	opp_table = dev_pm_opp_get_opp_table(dev);
-	if (IS_ERR(opp_table))
+
+	if (PTR_ERR(opp_table) == -EPROBE_DEFER)
 		return opp_table;
 
+	if (!opp_table)
+		return ERR_PTR(-ENOMEM);
+
 	/* Make sure there are no concurrent readers while updating opp_table */
 	WARN_ON(!list_empty(&opp_table->opp_list));
 
-- 
2.25.1

