From 481deb64fb44d63b0ebb8066daa55e227df98921 Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
Date: Thu, 24 Nov 2022 16:18:27 +0100
Subject: [PATCH] FIX GCC tools overwrite

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
---
Upstream-Status: Inappropriate [specific to ST boards]

 Makefile | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/Makefile b/Makefile
index 34d0076bb1..ed778f0a30 100644
--- a/Makefile
+++ b/Makefile
@@ -102,18 +102,18 @@ export Q ECHO
 # Toolchain
 ################################################################################
 
-HOSTCC			:=	gcc
+HOSTCC			?=	gcc
 export HOSTCC
 
-CC			:=	${CROSS_COMPILE}gcc
-CPP			:=	${CROSS_COMPILE}cpp
-AS			:=	${CROSS_COMPILE}gcc
-AR			:=	${CROSS_COMPILE}ar
-LINKER			:=	${CROSS_COMPILE}ld
-OC			:=	${CROSS_COMPILE}objcopy
-OD			:=	${CROSS_COMPILE}objdump
-NM			:=	${CROSS_COMPILE}nm
-PP			:=	${CROSS_COMPILE}gcc -E
+#CC			:=	${CROSS_COMPILE}gcc
+#CPP			:=	${CROSS_COMPILE}cpp
+#AS			:=	${CROSS_COMPILE}gcc
+#AR			:=	${CROSS_COMPILE}ar
+#LINKER			:=	${CROSS_COMPILE}ld
+#OC			:=	${CROSS_COMPILE}objcopy
+#OD			:=	${CROSS_COMPILE}objdump
+#NM			:=	${CROSS_COMPILE}nm
+#PP			:=	${CROSS_COMPILE}gcc -E
 DTC			:=	dtc
 
 # Use ${LD}.bfd instead if it exists (as absolute path or together with $PATH).
-- 
2.34.1

