From 31d0401abd304f934fda6f701cc11df2c145951f Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
Date: Fri, 2 Sep 2022 10:18:51 +0200
Subject: [PATCH 3/3] FIX openssl for nativesdk

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
---
 Makefile | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/Makefile b/Makefile
index f3b2098026..588bf5c1d8 100644
--- a/Makefile
+++ b/Makefile
@@ -179,18 +179,18 @@ endif
 # Toolchain
 ################################################################################
 
-HOSTCC			:=	gcc
+HOSTCC			?=	gcc
 export HOSTCC
 
-CC			:=	${CROSS_COMPILE}gcc
-CPP			:=	${CROSS_COMPILE}cpp
-AS			:=	${CROSS_COMPILE}gcc
-AR			:=	${CROSS_COMPILE}ar
-LINKER			:=	${CROSS_COMPILE}ld
-OC			:=	${CROSS_COMPILE}objcopy
-OD			:=	${CROSS_COMPILE}objdump
-NM			:=	${CROSS_COMPILE}nm
-PP			:=	${CROSS_COMPILE}gcc -E
+# CC			:=	${CROSS_COMPILE}gcc
+# CPP			:=	${CROSS_COMPILE}cpp
+# AS			:=	${CROSS_COMPILE}gcc
+# AR			:=	${CROSS_COMPILE}ar
+# LINKER			:=	${CROSS_COMPILE}ld
+# OC			:=	${CROSS_COMPILE}objcopy
+# OD			:=	${CROSS_COMPILE}objdump
+# NM			:=	${CROSS_COMPILE}nm
+# PP			:=	${CROSS_COMPILE}gcc -E
 DTC			:=	dtc
 
 # Use ${LD}.bfd instead if it exists (as absolute path or together with $PATH).
-- 
2.17.1

