From 10772b8cb5a78181da118ae229e67551023dc43c Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
Date: Tue, 12 May 2026 15:01:20 +0200
Subject: [PATCH 3/3] BINMAN: use importlib.resources instead of pkg_resources

Upstream-Status: Backport

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
---
 tools/binman/control.py | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/tools/binman/control.py b/tools/binman/control.py
index c6d3205b8c2..da97b2eb918 100644
--- a/tools/binman/control.py
+++ b/tools/binman/control.py
@@ -8,12 +8,12 @@
 from collections import OrderedDict
 import glob
 try:
-    import importlib.resources
-except ImportError:  # pragma: no cover
-    # for Python 3.6
+    import importlib.resources as importlib_resources
+    # for Python 3.6, 3.7 and 3.8
+    importlib_resources.files
+except (ImportError, AttributeError): # pragma: no cover
     import importlib_resources
 import os
-import pkg_resources
 import re
 
 import sys
@@ -96,8 +96,8 @@ def _ReadMissingBlobHelp():
             msg = ''
         return tag, msg
 
-    my_data = pkg_resources.resource_string(__name__, 'missing-blob-help')
-    re_tag = re.compile('^([-a-z0-9]+):$')
+    my_data = importlib_resources.files(__package__).joinpath('missing-blob-help').read_bytes()
+    re_tag = re.compile(r"^([-\.a-z0-9]+):$")
     result = {}
     tag = None
     msg = ''
@@ -151,8 +151,9 @@ def GetEntryModules(include_testing=True):
     Returns:
         Set of paths to entry class filenames
     """
-    glob_list = pkg_resources.resource_listdir(__name__, 'etype')
-    glob_list = [fname for fname in glob_list if fname.endswith('.py')]
+    entries = importlib_resources.files(__package__).joinpath('etype')
+    glob_list = [entry.name for entry in entries.iterdir()
+                 if entry.name.endswith('.py') and entry.is_file()]
     return set([os.path.splitext(os.path.basename(item))[0]
                 for item in glob_list
                 if include_testing or '_testing' not in item])
-- 
2.43.0

