From d3919a1412fba244357795faf4ed488149f33d5e Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
Date: Fri, 23 Feb 2024 11:56:29 +0100
Subject: [PATCH] MP2: silent_mode

Signed-off-by: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
---
 include/configs/stm32mp25_st_common.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/configs/stm32mp25_st_common.h b/include/configs/stm32mp25_st_common.h
index 3db57a2599..c56a40ec4f 100644
--- a/include/configs/stm32mp25_st_common.h
+++ b/include/configs/stm32mp25_st_common.h
@@ -9,6 +9,7 @@
 #define __CONFIG_STM32MP25_ST_COMMON_H__
 
 #define STM32MP_BOARD_EXTRA_ENV \
+	"silent=1\0" \
 	"usb_pgood_delay=1000\0" \
 	"console=ttySTM0\0"
 
-- 
2.34.1

