From d9d160739dd2ed9e1f59dacd1a0022d2ebf09280 Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@foss.st.com>
Date: Tue, 21 Jan 2025 11:58:26 +0100
Subject: [PATCH 2/2] TF-M-TEST by-pass git version check

---
Upstream-Status: Inappropriate [specific to ST boards]

 tests_reg/CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tests_reg/CMakeLists.txt b/tests_reg/CMakeLists.txt
index e93084a8..4220d050 100644
--- a/tests_reg/CMakeLists.txt
+++ b/tests_reg/CMakeLists.txt
@@ -25,7 +25,7 @@ include(${CONFIG_SPE_PATH}/cmake/spe_config.cmake)
 include(${CONFIG_SPE_PATH}/platform/config.cmake OPTIONAL)
 
 # Check current tf-m-tests is compatible to TF-M under verfication
-include(check_version)
+#include(check_version)
 
 include(${TFM_TOOLCHAIN_FILE})
 project(tfm_ns LANGUAGES C ASM)
-- 
2.34.1

